### PR TITLE
Switch to safer Open3 to determine OS_VERSION

### PR DESCRIPTION
**Pull request by no-itsbackpack**
Closes: https://github.com/googleapis/google-api-ruby-client/issues/880
References: 
 - article stating the problems with using backticks https://www.hilman.io/blog/2016/01/stop-using-backtick-to-run-shell-command-in-ruby/
 - discussion about deprecating backticks in ruby 3 (even though the proposal was rejected it is an interesting read) https://bugs.ruby-lang.org/issues/14385

This PR updates `version.rb` to use the safer `Open3` instead of backticks which is vulnerable to command injection when determining the `OS_VERSION`.

### CODE REVIEW

File: lib/google/apis/version.rb [hunk 1]
```diff
@@ -21,16 +23,19 @@ module Apis
     # @private
     OS_VERSION = begin
       if RUBY_PLATFORM =~ /mswin|win32|mingw|bccwin|cygwin/
-        `ver`.sub(/\s*\[Version\s*/, '/').sub(']', '')
+        output, _ = Open3.capture2('ver')
+        output.sub(/\s*\[Version\s*/, '/').sub(']', '')
       elsif RUBY_PLATFORM =~ /darwin/i
-        "Mac OS X/#{`sw_vers -productVersion`}"
+        output, _ = Open3.capture2('sw_vers -productVersion')
```
Comment (mistydemeo) [inline lib/google/apis/version.rb:29]: If it contains whitespace, the single-argument form will still get passed to a shell. You can avoid the shell by splitting it into separate arguments, in which case it gets `exec`ed directly. The documentation mentions how that works: https://ruby-doc.org/core-2.7.1/Process.html#method-c-spawn
```suggestion
        output, _ = Open3.capture2('sw_vers', '-productVersion')
```

File: lib/google/apis/version.rb [hunk 2]
```diff
@@ -21,16 +23,19 @@ module Apis
...
+        "Mac OS X/#{output}"
       elsif RUBY_PLATFORM == 'java'
         require 'java'
         name = java.lang.System.getProperty('os.name')
         version = java.lang.System.getProperty('os.version')
         "#{name}/#{version}"
       else
-        `uname -sr`.sub(' ', '/')
+        output, _ = Open3.capture2('uname -sr')
```
Comment (mistydemeo) [inline lib/google/apis/version.rb:37]:
```suggestion
        output, _ = Open3.capture2('uname', '-sr')
```

**Comment by no-itsbackpack** [comment]
@googlebot I signed it!

**Comment by no-itsbackpack** [comment]
👋🏾 any update on when we can get this merged? we are working on upgrading to the FCM v1 api and this is a blocker ❤️

**Comment by no-itsbackpack** [comment]
@googlebot I fixed it.

**Comment by mistydemeo** [comment]
@googlebot I fixed it.

**Comment by mistydemeo** [comment]
@googlebot I signed it.

**Comment by mistydemeo** [comment]
@googlebot I fixed it.

**Comment by no-itsbackpack** [comment]
@quartzmo thanks for the review ❤️, any updates on next steps for a merge 🙏🏾

**Comment by blowmage** [comment]
Is this something that should be generalized across more of the Google Ruby gems?

**Comment by no-itsbackpack** [comment]
@blowmage I'd vote for generalizing across more of the google ruby gems, only serves to make the gems more secure 🔒

**Comment by quartzmo** [comment]
Unfortunately I have no idea about when you can expect this to merge (and release). I would expect it to happen soon.

**Comment by dazuma** [comment]
The commands being executed here are all static; there's no interpolation going on. However, this shouldn't hurt anything, so we'll go ahead and merge it.

I can kick off an out of band release tonight. So you can expect a release with this patch in the next 24 hours.

@no-itsbackpack Would you mind linking to some outside reference regarding your preference against backticks, either from this PR or from #880 ?

**Comment by no-itsbackpack** [comment]
👋🏾 I just updated the PR body with a reference to the unintended side effects of using backticks and a discussion about deprecating backticks in ruby 3

**Comment by mistydemeo** [comment]
> The commands being executed here are all static; there's no interpolation going on.

Agreed, the risk is low. However, a shell *was* being invoked for the commands containing whitespace; my general thought is that it's best to execute as few subshells as possible. 😄

**Comment by dazuma** [comment]
Released: https://rubygems.org/gems/google-api-client/versions/0.39.2

**Pull request merged**